### DISCUSSION TITLE
hydration error when there is a chrome extension that modifies DOM

### DISCUSSION THREAD

**Discussion by jin2255** [remix-run/remix#4902] [Proposals]
### hydration error when there is a chrome extension that modifies DOM

I installed an extension that modifies DOM such as Grammar Checker into my browser and tried to develop a web app with Remix.
I get the following errors and the styles of the app were messed up.

_**Hydration failed because the initial UI does not match what was rendered on the server.
There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.**_

The important thing is that this error has already been mentioned many times and there is no exact solution
What's even more interesting is that the https://remix.run/ site doesn't have this error occur even if it was developed with Remix

I posted the error as above to https://github.com/remix-run/remix/issues/4822 and got a suggestion from [@machour](https://github.com/remix-run/remix/issues/4822#issuecomment-1350463059) that a discussion needs to happen.
Since I don't have a lot of experience in remix, I hope someone can help me with the exact solution to this problem.

As a solution, I also saw articles that found a way to wrap the content in a specific tag(e.g. `<div id="root"></div>`) and hydrate in that wrapper.
Will the remix be updated in this format in the future?

**Comment by ryanflorence** [comment]
This is a general React problem, not Remix specifically.

**Reply by machour** [reply]
@ryanflorence the difference is that remix hydrates the whole document, where other apps usually hydrate a `<div id="root"></div>`

We keep getting bug reports / questions about this warning, and I feel like we either want to reconsider what we hydrate, or rewrite this gotcha to provide all the answers to things folks have been wondering about: 
   https://remix.run/docs/en/v1/pages/gotchas#browser-extensions-injecting-code

Questions are:

- what are the implications of React switching to client rendering?
- is it a bug?
- is this a dev only issue?
- will it break my app in production for users with extensions?
- is it better to stick to React 17 for the time being?
- etc, etc

See also this (very correct) comment: https://github.com/remix-run/remix/pull/2682#issuecomment-1303797164

**Reply by lili21** [reply]
@machour so true. hydrating the whole document make the error more easier to happened.

**Reply by sergiodxa** [reply]
@machour I agree on updating the gotchas docs to explain more what's happening, specifically the part "will it break my app in production for users with extensions?"

**Reply by jin2255** [reply]
> This is a general React problem, not Remix specifically.

I know
I just wanted to know exact solution
I didn't tell it is Remix problem

**Reply by Ryongyon** [reply]
> This is a general React problem, not Remix specifically.

Expect to be more of a Chrome problem than a React problem 

In my research and observation, I have found that Chrome extensions that inject content into a page occasionally trigger this error, such as "React Developer Tools," which occasionally causes hydration errors. Later I read that Chrome extensions do not always inject content into a page at a fixed time, which means sometimes before the page is loaded, sometimes after the page is loaded. If you inject after the page is loaded, you won't get an error. If you inject before the page is loaded, you will naturally fail to hydrate. 

Once people bring up the subject, Remix officials will just say it's a React problem and do nothing about it. Why is Remix the only SSR framework with this problem? Or does Remix just not care about the developer's feelings? Remember that most browser extensions now inject scripted content into the page, and if this isn't fixed, it means Remix won't work in production. 

While we know that disabling Chrome extensions can fix this, we can't ask our users to disable Chrome extensions 

I believe the reason the Remix website doesn't have this issue is because they are using the React 17 version, if they are using the latest version they will have this issue

**Comment by lili21** [comment]
I created a similar discuss about this issue one month ago. https://github.com/remix-run/remix/discussions/4674

**Comment by jin2255** [comment]
for the more, I can't find any solution in this discusstion
they are telling me the reason of the error and telling me about it is React problem not remix
but I dont want to know about the reason, about it is react problem
I would like to get exact solution for this problem
but it seems nobody can help me for that
😭😭😭

**Reply by lili21** [reply]
You can try to wrap the unmatched part into Suspense, it can't fix the problem, but it can prevent the entire root switch to client rendering.

**Comment by Ryongyon** [comment]
Expect to be more of a Chrome problem than a React problem

In my research and observation, I have found that Chrome extensions that inject content into a page occasionally trigger this error, such as "React Developer Tools," which occasionally causes hydration errors. Later I read that Chrome extensions do not always inject content into a page at a fixed time, which means sometimes before the page is loaded, sometimes after the page is loaded. If you inject after the page is loaded, you won't get an error. If you inject before the page is loaded, you will naturally fail to hydrate.

Once people bring up the subject, Remix officials will just say it's a React problem and do nothing about it. Why is Remix the only SSR framework with this problem? Or does Remix just not care about the developer's feelings? Remember that most browser extensions now inject scripted content into the page, and if this isn't fixed, it means Remix won't work in production.

While we know that disabling Chrome extensions can fix this, we can't ask our users to disable Chrome extensions

I believe the reason the Remix website doesn't have this issue is because they are using the React 17 version, if they are using the latest version they will have this issue

**Reply by mdodge-ecgrow** [reply]
Same thing happens in Firefox when extensions are in use, so not just Chrome.

**Comment by ryanflorence** [comment]
Locked because this is not a proposal.

**Comment by kiliman** [comment]
I've got a solution for the hydration issues entirely in userland.

https://github.com/kiliman/remix-hydration-fix

**Comment by ryanflorence** [comment]
https://github.com/remix-run/remix/discussions/5244#discussioncomment-4888022

**Comment by kosiakMD** [comment]
that's odd, same isue

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
